From c5000fb7d8298f565970bc9833a6c82bcd2b8ac1 Mon Sep 17 00:00:00 2001
From: Paolo Bonzini <pbonzini@redhat.com>
Date: Wed, 21 Mar 2012 09:23:09 +0100
Subject: [PATCH 3/4] virtio-scsi: Fix virtio-scsi after cdb_is_read changes.

RH-Author: Paolo Bonzini <pbonzini@redhat.com>
Message-id: <1332321790-12045-4-git-send-email-pbonzini@redhat.com>
Patchwork-id: 38768
O-Subject: [RHEL6.3 PATCH seabios 3/4] virtio-scsi: Fix virtio-scsi after cdb_is_read changes.
Bugzilla: 801293
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>
RH-Acked-by: Gleb Natapov <gleb@redhat.com>
RH-Acked-by: Gerd Hoffmann <kraxel@redhat.com>

The previous patch changes the way TEST_UNIT_READY is composed in the
buffers and breaks virtio-scsi.

Reminder of the layout:

   request type  len     datain     0      1      2     in_num    out_num
---------------------------------------------------------------------------
           read  >0       true     req    resp   data      2         1
          write  >0       true     req    data   resp      1         2
test unit ready  =0      false*    req    resp             1         1

   * = true before patch 2, false now!  That's why out_num used to be
       computed first, while now in_num comes first.

Note that data_idx == in_num but that's just a curious coincidence.

Signed-off-by: Paolo Bonzini <pbonzini@redhat.com>
(cherry-picked from commit 8c313078917099a002d45f58d58ae2f4eb9a657f)
---
 src/virtio-scsi.c |   12 +++++++-----
 1 files changed, 7 insertions(+), 5 deletions(-)

Signed-off-by: Michal Novotny <minovotn@redhat.com>
---
 src/virtio-scsi.c |   12 +++++++-----
 1 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/src/virtio-scsi.c b/src/virtio-scsi.c
index 76c5f29..0aa3388 100644
--- a/src/virtio-scsi.c
+++ b/src/virtio-scsi.c
@@ -46,9 +46,8 @@ virtio_scsi_cmd(u16 ioaddr, struct vring_virtqueue *vq, struct disk_op_s *op,
 
     u32 len = op->count * blocksize;
     int datain = cdb_is_read(cdbcmd, blocksize);
-    int data_idx = (datain ? 2 : 1);
-    int out_num = (datain ? 1 : 2);
-    int in_num = (len ? 3 : 2) - out_num;
+    int in_num = (datain ? 2 : 1);
+    int out_num = (len ? 3 : 2) - in_num;
 
     sg[0].addr   = MAKE_FLATPTR(GET_SEG(SS), &req);
     sg[0].length = sizeof(req);
@@ -56,8 +55,11 @@ virtio_scsi_cmd(u16 ioaddr, struct vring_virtqueue *vq, struct disk_op_s *op,
     sg[out_num].addr   = MAKE_FLATPTR(GET_SEG(SS), &resp);
     sg[out_num].length = sizeof(resp);
 
-    sg[data_idx].addr   = op->buf_fl;
-    sg[data_idx].length = len;
+    if (len) {
+        int data_idx = (datain ? 2 : 1);
+        sg[data_idx].addr   = op->buf_fl;
+        sg[data_idx].length = len;
+    }
 
     /* Add to virtqueue and kick host */
     vring_add_buf(vq, sg, out_num, in_num, 0, 0);
-- 
1.7.7.6

